### PR TITLE
DRAFT - Feature/339 METAR Parsing

### LINKED ISSUE #339: Ability for OpenDCS to read METAR files and parse into readable formats
**Is your feature request related to a problem? Please describe.**
The abstract web pull feature in OpenDCS allows you to read human-readable data and parse it into CWMS time series, but to my knowledge is not able to parse something as complex as a METAR message.  I am also not sure whether OpenDCS allows you to pull data from an LDM.

**Describe the solution you'd like**
Request the ability to read METAR messages and decode them.

**Describe alternatives you've considered**
This data ingest process is currently done with a METAR decoder java program written by @hoffmanjs, which reads the METAR files from the LDM, runs them through a basic screening process, then parses them into human-readable shef files and pushes them through processShefit to pull the data into the Oracle database.  Our understanding from what we've been told about the cloud migration is that these jobs cannot be there in CWBI, and we need to look at alternative ways to get the data.

**Additional context**
METAR (meteorological data stations from NOAA) data may also be able to be pulled through an abstract web source if ingesting LDM data isn't possible in OpenDCS.  More research would need to be done on this.  A sample METAR message and it's decoded data follows.  @hoffmanjs can also provide NOAA's documentation on how to decode METAR messages, which is quite extensive due to the types of data that could be in a message.  We require 10 parameters from these messages to run our reservoir evaporation program.

**Sample Message**
KBIS 181252Z 32018G26KT 10SM -RA BKN016 BKN021 OVC080 10/08 A2972 RMK AO2 PK WND 29027/1203 RAE05B43 SLP065 P0001 T01000083

**Decoded data**
METAR for KBIS (Bismarck Arpt, ND, US) observed at 1252 UTC 18 Oct 2023
  Text: KBIS 181252Z 32018G26KT 10SM -RA BKN016 BKN021 OVC080 10/08 A2972 RMK AO2 PK WND 29027/1203 RAE05B43 SLP065 P0001 T01000083
  Temperature: 10C (50.0 F)
  Dewpoint: 8.3C (46.9 F)
  Altimeter: 29.72 inches Hg (1006.5 mb)
  Sea level pressure: 1006.5 mb
  Winds: from 320 degrees at 18 knots gusting to 26 knots
  Visibility: 10+ sm
  Ceiling: 1600 feet AGL
  Clouds: broken clouds at 1600 feet AGL, broken clouds at 2100 feet AGL, overcast cloud deck at 8000 feet AGL
  Weather: -RA (light rain)

### PR DESCRIPTION
**Pull request by ktarbet**
## Problem Description

Fixes #339 

## Solution

use existing METAR parser: https://github.com/mivek/MetarParser

## how you tested the change

Describe what was done to test the change. This section can be left blank 
if automated tests demonstrating usage are provided in the PR.

## Where the following done:

- [ ] Tests. Check all that apply:
   - [ ] Unit tests created or modified that run during ant test.
   - [ ] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [ ] Were relevant config element (e.g. XML data) updated as appropriate

If you aren't sure leave unchecked and we will help guide you to want needs changing where.

### CODE REVIEW

**Comment by ktarbet** [comment]
@MikeNeilson  We are not sure about adding the dependency.  We see this on sonatype.com...

https://central.sonatype.com/artifact/io.github.mivek/metarParser?smo=true

**Comment by MikeNeilson** [comment]
What aren't you sure about? Sonartype reports no known vulnerabilities, depedencies are minimal and ones we already include.

Appears to track correct to the github project owner and has recent releases.

File: src/main/java/decodes/decoder/MetarProcess.java [hunk 1]
```diff
@@ -0,0 +1,729 @@
...
+		// times.
+		debugSdf.setTimeZone(TimeZone.getTimeZone(tzID));
+		msg.getTimer().setTimeZoneName(tzID);
+		msg.getTimer().setComplete(msg.getMessageTime());
+		cal = msg.getTimer().getCalendar();
+
+		qualityFlags = shefQuality2Flags('Z'); // default is no qual-checks performed
+		
+		String msgData = new String(rawmsg.getData());
+		String shefdata = msgData.substring(rawmsg.getHeaderLength());
```
Comment (MikeNeilson) [inline src/main/java/decodes/decoder/MetarProcess.java:131]: shef? I assume you copied this from ShefParser and just haven't updated this method yet?

Comment (ktarbet) [reply src/main/java/decodes/decoder/MetarProcess.java:131]: Yes, just starting with a copy of ShefParser.

**Comment by ktarbet** [comment]
I found a file  pom.template.   That appears where we could all more maven repositories?  I haven't succeeded yet in adding the Metar parser dependency.

**Comment by MikeNeilson** [comment]
Oh I see, it looks like it may not have been uploaded correctly. ivysettings.xml contains the artifact repositories. But I don't think that will help.

File: ivy.xml [hunk 2]
```diff
@@ -84,6 +84,7 @@
         <dependency org="org.jdbi" name="jdbi3-sqlobject" rev="3.38.2"/>
         <dependency org="org.jdbi" name="jdbi3-postgres" rev="3.38.2"/>
         <dependency org="org.flywaydb" name="flyway-core" rev="9.8.1"/>
+        <dependency org="io.github.mivek" name="metarParser" rev="2.16.0"/>
```
Comment (MikeNeilson) [inline ivy.xml:87]: `metarParser-services`  is the correct dependency per the projects readme.

**Comment by ktarbet** [comment]
I'm getting an error because I'm using Java 1.8 -- the OpenDCS default;  but MetarService  is compiled for  Java 11.

```
import io.github.mivek.service.MetarService;
    [javac]                               ^
    [javac]   bad class file: C:\Users\?\.ant\cache\io.github.mivek\metarParser-services\jars\metarParser-services-2.16.0.jar(/io/github/mivek/service/MetarService.class)
    [javac]     class file has wrong version 61.0, should be 55.0

**Comment by MikeNeilson** [comment]
> I'm getting an error because I'm using Java 1.8 -- the OpenDCS default; but MetarService is compiled for Java 11.
> 
> ```
> import io.github.mivek.service.MetarService;
>     [javac]                               ^
>     [javac]   bad class file: C:\Users\?\.ant\cache\io.github.mivek\metarParser-services\jars\metarParser-services-2.16.0.jar(/io/github/mivek/service/MetarService.class)
>     [javac]     class file has wrong version 61.0, should be 55.0
> ```


That's unfortunate. We could fork it and get it working in java 8, but might be easier to just write a simple parser.

**Comment by ktarbet** [comment]
Are we restricting 7.0.*  versions of OpenDCS to java 1.8?

I know having a zip file output is needed before we drop izpack (which needs Java 1.8).  If that is the only hurdle to dropping java 1.8?    

I'm willing to work on the zip file:  https://github.com/opendcs/opendcs/issues/578

**Comment by MikeNeilson** [comment]
Unfortunately that was only keeping the build and tests from working in over than 1.8. 1.8 is simply the current target for 7 as the community asked for that. We'd have to coordinate with several large users. I'm fairly confident USACE would be fine, I'm not sure where USBR is with java 11 (@agilmore2 ).

That said, a multi version jar, where is in a java11 folder would perhaps be reasonable and we do support java 11+. Might be upsetting to those who can't upgrade yet, but nothing is perfect.

We can bring this up with the community again at the in-person but 7.0 is currently requiring to run fully on java 1.8 due to community requirements.

**Pull request closed without merging**